libfuzzer/libfuzz_harness.cc

```diff
@@ -95,6 +341,146 @@ ubpf_vm_ptr create_ubpf_vm(const std::vector<uint8_t>& program_code)
     return vm;
 }
 
+/**
+ * @brief Classify the given register value as packet, context, stack, map, or unknown.
+ */
+typedef enum class _address_type {
+    Packet,
+    Context,
+    Stack,
+    Map,
+    Unknown
+} address_type_t;
+
+/**
+ * @brief Given a register value, classify it as packet, context, stack, or unknown.
+ *
+ * @param[in] context Pointer to the context structure.
+ * @param[in] register_value Register value to classify.
+ * @retval address_type_t::Packet The register value is within the packet data.
+ * @retval address_type_t::Context The register value is within the context structure.
+ * @retval address_type_t::Stack The register value is within the stack.
+ * @retval address_type_t::Unknown The register value is unknown.
+ */
+address_type_t ubpf_classify_address(ubpf_context_t* context, uint64_t register_value)
+{
+    uintptr_t register_value_ptr = reinterpret_cast<uintptr_t>(register_value);
+    uintptr_t stack_start = reinterpret_cast<uintptr_t>(context->stack_start);
+    uintptr_t stack_end = reinterpret_cast<uintptr_t>(context->stack_end);
+    uintptr_t context_start = reinterpret_cast<uintptr_t>(context);
+    uintptr_t context_end = context_start + sizeof(ubpf_context_t);
+    uintptr_t packet_start = reinterpret_cast<uintptr_t>(context->data);
+    uintptr_t packet_end = reinterpret_cast<uintptr_t>(context->data_end);
+
+    if (register_value >= stack_start && register_value < stack_end) {
+        return address_type_t::Stack;
+    } else if (register_value >= context_start && register_value < context_end) {
+        return address_type_t::Context;
+    } else if (register_value >= packet_start && register_value < packet_end) {
+        return address_type_t::Packet;
+    } else {
+        return address_type_t::Unknown;
+    }
+}
+
+/**
+ * @brief Function invoked prior to executing each instruction in the program.
+ *
+ * @param[in] context Context passed to the program.
+ * @param[in] program_counter The program counter (the index of the instruction to execute)
+ * @param[in] registers The register values.
+ * @param[in] stack_start The start of the stack.
+ * @param[in] stack_length The length of the stack.
+ * @param[in] register_mask The set of registers that have been modified since the start of the program.
+ * @param[in] stack_mask The set of stack locations that have been modified since the start of the program.
+ */
+void
+ubpf_debug_function(
+    void* context, int program_counter, const uint64_t registers[16], const uint8_t* stack_start, size_t stack_length, uint64_t register_mask, const uint8_t* stack_mask)
+{
+#if 1
```

Comment:
clean this up

Reply:
Switched this to a CMake option for now. Once https://github.com/vbpf/ebpf-verifier/pull/729 is merged we can decide if we want this to be always on.